File: python/euler/bin/p0024.py
# project euler: problem 24

from collections.abc import MutableSequence
from math import factorial


def make_fact_tbl(num: int, depth: int) -> list[int]:
    result = []
    i = num - 1
    divisor = factorial(num - depth)
    for _ in range(depth):
        result.append(factorial(i) // divisor)
        i -= 1

    return result


def compute(nth: int, elm_lst: MutableSequence[int], depth: int) -> str:
    idx = nth - 1
    result = []
    for n in make_fact_tbl(len(elm_lst), depth):
        result.append(elm_lst.pop(idx // n))
        idx %= n

    return ''.join(map(str, result))


def solve() -> str:
    lst = list(range(10))
    return compute(1_000_000, lst, len(lst))

File: typescript/euler/bin/p0024.ts
// project euler: problem 24

import { factorial } from "../lib/math.ts";
import { range } from "../lib/util.ts";

const trunc = Math.trunc;

function makeFactTbl(n: number, depth: number): number[] {
  const result: number[] = [];
  let i = n - 1;
  const divisor = factorial(n - depth);
  for (const _ of range(0, depth)) {
    result.push(trunc(factorial(i) / divisor));
    i -= 1;
  }

  return result;
}

export const compute = (nth: number, elmLst: number[]): string => {
  const depth = elmLst.length;
  let idx = nth - 1;
  const result: number[] = [];
  for (const n of makeFactTbl(elmLst.length, depth)) {
    result.push(elmLst[trunc(idx / n)]);
    elmLst.splice(trunc(idx / n), 1);
    idx %= n;
  }

  return result.map((x) => String(x)).join("");
};

export const solve = (): string => compute(1_000_000, range(0, 10));
